rpc/grpc/api.go

```diff
@@ -37,3 +51,372 @@ func (bapi *broadcastAPI) BroadcastTx(ctx context.Context, req *RequestBroadcast
 		},
 	}, nil
 }
+
+type BlockAPI struct {
+	sync.Mutex
+	heightListeners      map[chan NewHeightEvent]struct{}
+	newBlockSubscription eventstypes.Subscription
+	subscriptionID       string
+	subscriptionQuery    pubsub.Query
+}
+
+func NewBlockAPI() *BlockAPI {
+	return &BlockAPI{
+		heightListeners:   make(map[chan NewHeightEvent]struct{}, 1000),
+		subscriptionID:    fmt.Sprintf("block-api-subscription-%s", rand.Str(6)),
+		subscriptionQuery: eventstypes.EventQueryNewBlock,
+	}
+}
+
+func (blockAPI *BlockAPI) StartNewBlockEventListener(ctx context.Context) error {
+	env := core.GetEnvironment()
+	if blockAPI.newBlockSubscription == nil {
+		var err error
+		blockAPI.newBlockSubscription, err = env.EventBus.Subscribe(
+			ctx,
+			blockAPI.subscriptionID,
+			blockAPI.subscriptionQuery,
+			500,
+		)
+		if err != nil {
+			env.Logger.Error("Failed to subscribe to new blocks", "err", err)
+			return err
+		}
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-blockAPI.newBlockSubscription.Cancelled():
+			env.Logger.Error("cancelled grpc subscription. retrying")
+			ok, err := blockAPI.retryNewBlocksSubscription(ctx)
+			if err != nil {
+				return err
+			}
+			if !ok {
+				// this will happen when the context is done. we can stop here
+				return nil
+			}
+		case event, ok := <-blockAPI.newBlockSubscription.Out():
+			if !ok {
+				env.Logger.Error("new blocks subscription closed. re-subscribing")
+				ok, err := blockAPI.retryNewBlocksSubscription(ctx)
+				if err != nil {
+					return err
+				}
+				if !ok {
+					// this will happen when the context is done. we can stop here
+					return nil
+				}
+				continue
+			}
+			newBlockEvent, ok := event.Events()[eventstypes.EventTypeKey]
+			if !ok || len(newBlockEvent) == 0 || newBlockEvent[0] != eventstypes.EventNewBlock {
+				continue
+			}
+			data, ok := event.Data().(eventstypes.EventDataNewBlock)
+			if !ok {
+				env.Logger.Debug("couldn't cast event data to new block")
+				return fmt.Errorf("couldn't cast event data to new block. Events: %s", event.Events())
+			}
+			blockAPI.broadcastToListeners(ctx, data.Block.Height, data.Block.Hash())
+		}
+	}
+}
+
+// RetryAttempts the number of retry times when the subscription is closed.
+const RetryAttempts = 6
+
+// SubscriptionCapacity the maximum number of pending blocks in the subscription.
+const SubscriptionCapacity = 500
+
+func (blockAPI *BlockAPI) retryNewBlocksSubscription(ctx context.Context) (bool, error) {
```

Comment:
[question][nit] what does the bool return argument indicate? It seems like `(isSuccess bool, err error)` IMO the bool argument is unnecessary b/c either this function returns:
1. true, no error
2. false, an error

So the bool argument could be removed and this could just return an error if the retry failed or nil if the retry succeeded